add basic climate support

Schluggi opened the pull request:
<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Description](#description)
- [License Acceptance](#license-acceptance)
- [Type of changes](#type-of-changes)
- [Checklist](#checklist)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<!--- Provide a general summary of your changes in the Title above -->

# Description
I added basic support for climate (#222). This is not done yet and I need some help.

In my PR I add a new class called `ClimateSubscriber` to handle all the special topics needed for a `climate` entity. I'm pretty sure this not the supposed way to do this. Moreover there is a lot of code duplication in `ClimateSubscriber` and `Subscriber`. `climate` is the first entity with multiple topics as far as I can see. Furthermore it would be much nicer to have different callbacks for each command. Therefore the `Subscriber` class must be rewritten. 

Supported features so far:
- Modes
- Target temperature
- Current temperature

Whats missing:
- Actions
- Humidity
- Fan
- Power
- Presets
- Swing
- Temperature High/Low

<!--- Describe your changes in detail -->

# License Acceptance

- [x] This repository is Apache version 2.0 licensed and by making this PR, I am contributing my changes to the repository under the terms of the Apache 2 license.

# Type of changes

<!--- What types of changes does your submission introduce? Put an `x` in all the boxes that apply: [x] -->

- [ ] Add/update a helper script
- [ ] Add/update link to an external resource like a blog post or video
- [ ] Bug fix
- [x] New feature
- [ ] Test updates
- [ ] Text cleanups/updates
- [ ] New release to PyPi

# Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. [x] -->
<!--- If you're unsure about any of these, don't hesitate to ask. I'm happy to help! -->

- [x] I have read the [CONTRIBUTING](https://github.com/unixorn/ha-mqtt-discovery/blob/main/Contributing.md) document.
- [x] All new and existing tests pass.
- [ ] Any scripts added use `#!/usr/bin/env interpreter` instead of potentially platform-specific direct paths (`#!/bin/sh` is an allowed exception)
- [ ] Scripts added/updated in this PR are all marked executable.
- [ ] Scripts added/updated in this PR _do not_ have a language file extension unless they are meant to be sourced and not run standalone. No one should have to know if a script was written in bash, python, ruby or whatever. Not including file extensions makes it easier to rewrite the script in another language later without having to change every reference to the previous version.
- [ ] I have confirmed that any links added or updated in my PR are valid.

Schluggi committed:
add support for climate

kratz00 committed:
Update ruff to version 0.12.3

Signed-off-by: Steffen Pankratz <kratz00@gmx.de>

Signed-off-by: Schluggi <github@das-it-gesicht.de>

kratz00 committed:
Update pyaml to version 25.7.0

Signed-off-by: Steffen Pankratz <kratz00@gmx.de>

Signed-off-by: Schluggi <github@das-it-gesicht.de>

fix README

Signed-off-by: Schluggi <github@das-it-gesicht.de>

[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Signed-off-by: Schluggi <github@das-it-gesicht.de>

unixorn committed:
Ask how contributors want to be credited

Ask for contributor mastodon account so we can credit them properly in
release announcements. Also ask if there's a different way they'd prefer
to be credited.

Signed-off-by: Joe Block <jpb@unixorn.net>
Signed-off-by: Schluggi <github@das-it-gesicht.de>

kratz00 committed:
Update ruff to version 0.12.4

Signed-off-by: Steffen Pankratz <kratz00@gmx.de>

Signed-off-by: Schluggi <github@das-it-gesicht.de>

kratz00 committed:
Update ruff to version 0.12.5

Signed-off-by: Steffen Pankratz <kratz00@gmx.de>

Signed-off-by: Schluggi <github@das-it-gesicht.de>

kratz00 committed:
Update ruff to version 0.12.7

Signed-off-by: Steffen Pankratz <kratz00@gmx.de>

Signed-off-by: Schluggi <github@das-it-gesicht.de>

kratz00 commented:
How about?

- class _Subscriber_: replace *command_callback* with a dictionary (key: name of command topic, value: tuple(topic name, callback))
    
 This class can then be used to create something like *SimpleSubscriber* which only covers the classical *command_topic* or something more complex like your *ClimateSubscriber* class. If this makes sense to you, I can create a feature branch with the proposed changes.

Schluggi commented:
@kratz00 
Please let me know, when it is ready.

PS: The DSO check is false positive I think: [e4be520](https://github.com/unixorn/ha-mqtt-discoverable/pull/389/commits/e4be52002e2509e03d17e0a51c2baa4013d5e4ac) and [e3fe2bf](https://github.com/unixorn/ha-mqtt-discoverable/pull/389/commits/e3fe2bf4af01358543f4d6e9c510b89e7a437147) are shown as `verified` in github.

Schluggi committed:
add support for climate

add support for climate

Signed-off-by: Schluggi <github@das-it-gesicht.de>

 This is the commit message #3:

add support for climate

Signed-off-by: Schluggi <github@das-it-gesicht.de>

 This is the commit message #4:

add support for climate

Signed-off-by: Schluggi <github@das-it-gesicht.de>

add support for climate

Signed-off-by: Schluggi <github@das-it-gesicht.de>

fix pytest

jvanoosterhout commented:
@kratz00 @Schluggi I was wondering how far you came with this issue. I would also like this package to support multi topic devices (off topic: covers also rely on more than one topics). Locking briefly at the code I thought of the same idea as @kratz00. While searching for related issues and figuring out how to contribute I found this Issue. Let me know if/how I could support to rewrite the subscriber, though I guess also the _update_state must be rewritten.

Best regards, Jeroen

Schluggi commented:
Hi @jvanoosterhout, since I never got a reply from @kratz00, this PR is still pending. And at least for my initial usecase totally outdated. Therefore I will close it.

kratz00 commented:
@Schluggi Sorry, what question didn't I reply to?

@jvanoosterhout Thanks for asking, at the moment it's covered, am working on refactoring things to include climate support and also #522 - at the same time I am also trying to separate all the MQTT client handling from the entities. As with so many open source projects "It's done when it's done" :smile:
